system-addon/lib/FaviconFeed.jsm

```diff
@@ -10,23 +10,122 @@ Cu.importGlobalProperties(["fetch", "URL"]);
 const TIPPYTOP_UPDATE_TIME = ONE_DAY;
 const TIPPYTOP_RETRY_DELAY = FIVE_MINUTES;
+const MIN_FAVICON_SIZE = 96;
+
+/**
+ * Get favicon info (uri and size) for a uri from Places.
+ *
+ * @param uri {nsIURI} Page to check for favicon data
+ * @returns A promise of an object (possibly null) containing the data
+ */
+function getFaviconInfo(uri) {
+  // Use 0 to get the biggest width available
+  const preferredWidth = 0;
+  return new Promise(resolve => PlacesUtils.favicons.getFaviconDataForPage(
+    uri,
+    // Package up the icon data in an object if we have it; otherwise null
+    (iconUri, faviconLength, favicon, mimeType, faviconSize) =>
+      resolve(iconUri ? {iconUri, faviconSize} : null),
+    preferredWidth));
+}
+
+/**
+ * Fetches visit paths for a given URL from its most recent visit in Places.
+ *
+ * Note that this includes the URL itself as well as all the following
+ * permenent&temporary redirected URLs if any.
+ *
+ * @param {String} a URL string
+ *
+ * @returns {Array} Returns an array containing objects as
+ *   {int}    visit_id: ID of the visit in moz_historyvisits.
+ *   {String} url: URL of the redirected URL.
+ */
+async function fetchVisitPaths(url) {
+  const query = `
+    WITH RECURSIVE path(visit_id)
+    AS (
+      SELECT v.id
+      FROM moz_places h
+      JOIN moz_historyvisits v
+        ON v.place_id = h.id
+      WHERE h.url_hash = hash(:url) AND h.url = :url
+        AND v.visit_date = h.last_visit_date
+
+      UNION
+
+      SELECT id
+      FROM moz_historyvisits
+      JOIN path
+        ON visit_id = from_visit
+      WHERE visit_type IN
+        (${PlacesUtils.history.TRANSITIONS.REDIRECT_PERMANENT},
+         ${PlacesUtils.history.TRANSITIONS.REDIRECT_TEMPORARY})
+    )
+    SELECT visit_id, (
+      SELECT (
+        SELECT url
+        FROM moz_places
+        WHERE id = place_id)
+      FROM moz_historyvisits
+      WHERE id = visit_id) AS url
+    FROM path
+  `;
+
+  const visits = await NewTabUtils.activityStreamProvider.executePlacesQuery(query, {
+    columns: ["visit_id", "url"],
+    params: {url}
+  });
+  return visits;
+}
+
+/**
+ * Fetch favicon for a url by following its redirects in Places.
+ *
+ * This can improve the rich icon coverage for Top Sites since Places only
+ * associates the favicon to the final url if the original one gets redirected.
+ * Note this is not an urgent request, hence it is dispatched to the main
+ * thread idle handler to avoid any possible performance impact.
+ */
+async function fetchIconFromRedirects(url) {
+  const visitPaths = await fetchVisitPaths(url);
+  if (visitPaths.length > 1) {
+    const lastVisit = visitPaths.pop();
+    const redirectedUri = Services.io.newURI(lastVisit.url);
+    const iconInfo = await getFaviconInfo(redirectedUri);
+    if (iconInfo && iconInfo.faviconSize >= MIN_FAVICON_SIZE) {
+      PlacesUtils.favicons.setAndFetchFaviconForPage(
+        Services.io.newURI(url),
+        iconInfo.iconUri,
```

Comment:
Does this use the original icon uri? I would have expected something like tippytop's ` iconUri = iconUri.mutate().setRef("tippytop").finalize();`

Reply:
You mean, something like `.setRef("redirect").finalize();`?

I thought that if we'd only use that ref to identify the favicon type in telemetry (as it's commented [here](https://github.com/mozilla/activity-stream/blob/master/system-addon/lib/FaviconFeed.jsm#L233)), then this really is just a regular rich icon.